api/filrewardsd/service/service.go

```diff
@@ -0,0 +1,379 @@
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/textileio/go-threads/util"
+	analytics "github.com/textileio/textile/v2/api/analyticsd/client"
+	analyticspb "github.com/textileio/textile/v2/api/analyticsd/pb"
+	pb "github.com/textileio/textile/v2/api/filrewardsd/pb"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+var log = logging.Logger("filrewards")
+
+var rewardMeta = map[pb.Reward]meta{
+	pb.Reward_REWARD_FIRST_KEY_ACCOUNT_CREATED:    {factor: 3},
+	pb.Reward_REWARD_FIRST_KEY_USER_CREATED:       {factor: 1},
+	pb.Reward_REWARD_FIRST_ORG_CREATED:            {factor: 3},
+	pb.Reward_REWARD_INITIAL_BILLING_SETUP:        {factor: 1},
+	pb.Reward_REWARD_FIRST_BUCKET_CREATED:         {factor: 2},
+	pb.Reward_REWARD_FIRST_BUCKET_ARCHIVE_CREATED: {factor: 2},
+	pb.Reward_REWARD_FIRST_MAILBOX_CREATED:        {factor: 1},
+	pb.Reward_REWARD_FIRST_THREAD_DB_CREATED:      {factor: 1},
+}
+
+type meta struct {
+	factor int32
+}
+
+type rewardRecord struct {
+	Key               string                  `bson:"key"`
+	AccountType       analyticspb.AccountType `bson:"account_type"`
+	Reward            pb.Reward               `bson:"reward"`
+	Factor            int32                   `bson:"factor"`
+	BaseAttoFILReward int32                   `bson:"base_atto_fil_reward"`
+	CreatedAt         time.Time               `bson:"created_at"`
+	ClaimedAt         *time.Time              `bson:"claimed_at"`
+}
+
+var _ pb.FilRewardsServiceServer = (*Service)(nil)
+
+type Service struct {
+	col               *mongo.Collection
+	ac                *analytics.Client
+	rewardsCache      map[string]map[pb.Reward]struct{}
+	baseAttoFILReward int32
+	server            *grpc.Server
+}
+
+type Config struct {
+	Listener            net.Listener
```

Comment:
Notice how the listener is abstracted one level up out of the gRPC service. This is because I found a great listener you can use for testing that allows you to run your client and server in the same process using an in memory listener/connection. Great for debugging and tests in general. See `service_test.go` for this type of use.